pkg/compact/planner.go

```diff
@@ -155,28 +161,62 @@ func selectMetas(ranges []int64, noCompactMarked map[ulid.ULID]*metadata.NoCompa
 // selectOverlappingMetas returns all dirs with overlapping time ranges.
 // It expects sorted input by mint and returns the overlapping dirs in the same order as received.
 // Copied and adjusted from https://github.com/prometheus/prometheus/blob/3d8826a3d42566684283a9b7f7e812e412c24407/tsdb/compact.go#L268.
-func selectOverlappingMetas(metasByMinTime []*metadata.Meta) []*metadata.Meta {
+func selectOverlappingMetas(metasByMinTime []*metadata.Meta) []CompactionTask {
 	if len(metasByMinTime) < 2 {
 		return nil
 	}
-	var overlappingMetas []*metadata.Meta
-	globalMaxt := metasByMinTime[0].MaxTime
-	for i, m := range metasByMinTime[1:] {
-		if m.MinTime < globalMaxt {
-			if len(overlappingMetas) == 0 {
-				// When it is the first overlap, need to add the last one as well.
-				overlappingMetas = append(overlappingMetas, metasByMinTime[i])
+
+	groups := make([][]*metadata.Meta, len(metasByMinTime))
+	for i, m := range metasByMinTime {
+		groups[i] = []*metadata.Meta{m}
+	}
+
+	// Iterate through all metas and merge overlapping blocks into groups.
+	// We compare each block's max time with the next block's min time.
+	// If we detect an overlap, we merge all blocks from the current block's group
+	// into the group of the next block.
+	// We also adjust the head of the next group's head (zero-th element) to be the block with the
+	// highest max-time. This allows groups to be used as heaps, and helps us detect
+	// overlaps when the next block is contained in the current block.
+	// See test case https://github.com/thanos-io/thanos/blob/04106d7a7add7f47025c00422c80f746650c1b97/pkg/compact/planner_test.go#L310-L321.
+loopMetas:
+	for i := range metasByMinTime {
```

Comment:
Critical code: Reviewers please be careful here @thanos-io/thanos-maintainers - the bug here can cause irreversible data malformations